### PR TITLE
sFlow High Level Design

### PR DESCRIPTION
**Pull request by padmanarayana**
This document describes the high level design of sFlow in SONiC.

### CODE REVIEW

File: doc/sflow/sflow_hld.md [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+3. sFlow counter support needed.
+4. Polling interval for sFlow counter.
+
+### **Phase III:**
+1. sFlow extended switch support.
+2. sFlow extended router support.
+
+### Not planned to be supported:
+1. sFlow enable/disable option interface level as well as global level.
+2. sFlow configuration at interface level.
```
Comment (lguohan) [inline doc/sflow/sflow_hld.md]: why do we have such limitation?

File: doc/sflow/sflow_hld.md [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+genetlink_group                       = 1*4DIGIT                           ; genetlink multicast group id
+```
+
+A new SFLOW_TABLE is added to the AppDB:
+
+```
+; Defines schema for SFLOW_TABLE table which contains parameters that need to be set for sFlow
+key                 = SFLOW_TABLE:sflow
+; field             = value
+sampling_rate       = 1*5DIGIT           ; 1 / (sflow sampling rate as a percentage of bandwidth)
```
Comment (lguohan) [inline doc/sflow/sflow_hld.md]: this is unclear to me. what does 1 mean here? does it mean we sample every packet?

File: doc/sflow/sflow_hld.md [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+	2. 	Management port
+5. Set the default sampling rate to 0, configuring non-default value should enable sFlow and start sampling.
+6. Default sFlow sample size should be set to 128 bytes.
+7. Support sFlow related
+	1. CLI show/config commands
+	2. syslogs
+
+
+### **Phase II:**
+1. sFlow should be supported on portchannel interfaces.
```
Comment (madhupalu) [inline doc/sflow/sflow_hld.md]: do you support break-out interfaces? Is there any hsflowd limitation to support breakouts?

File: doc/sflow/sflow_hld.md [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+## 7 **sFlow in Virtual Switch**
+
+On the SONiC VS, SAI calls would map to the tc_sample commands on the switchport interfaces (http://man7.org/linux/man-pages/man8/tc-sample.8.html).
+
+## 8 **Build**
+
+* The host-sflow package will be built with the mod_sonic callback implementations using the FEATURES="SONIC" option
+
+## 9 **Restrictions**
+* /etc/hsflowd.conf should not be modified manually. While it should be possible to change /etc/hsflowd.conf manually and restart the sflow container, it is not recommended.
```
Comment (madhupalu) [inline doc/sflow/sflow_hld.md]: is there any hsflowd/InMon license implications? Community interested to learn on the implications if any.

File: doc/sflow/sflow_hld.md [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+  collector { ip = 10.144.1.2 UDPPort=6344 }
+  datagramBytes = 1500
+}
+```
+
+sflowmgrd also listens to SFLOW to propogate the sampling rate changes to App DB SFLOW_TABLE.
+
+#### hsflowd service
+
+hsflowd provides an module adaptation layer for interfacing with the NOS. In the host-sflow repo, a src/Linux/mod_sonic.c adaption layer will be provided for hsflowd APIs to SONiC that deal with hsflowd initialization, configuration changes, packet sample consumption etc. More specifically, SONiC will register and provide callbacks for the following HSP events:
```
Comment (madhupalu) [inline doc/sflow/sflow_hld.md]: What is the configurations need to expose hsflowd/InMon ? If user wanted to expose some custom configuration such as sampling rate what is the strategy, where do maintain the custom configurations? is it in SONiC Repo or hsflowd/InMon Repos?

File: doc/sflow/sflow_hld.md (hunk 1 above)
Comment (padmanarayana) [reply doc/sflow/sflow_hld.md]: Updated HLD. This will be supported.

File: doc/sflow/sflow_hld.md (hunk 2 above)
Comment (padmanarayana) [reply doc/sflow/sflow_hld.md]: Update definition of sampling rate (it is as per the SAI definition).

File: doc/sflow/sflow_hld.md (hunk 3 above)
Comment (padmanarayana) [reply doc/sflow/sflow_hld.md]: break-out interfaces are not currently supported.

File: doc/sflow/sflow_hld.md (hunk 4 above)
Comment (padmanarayana) [reply doc/sflow/sflow_hld.md]: https://github.com/sflow/host-sflow/blob/master/LICENSE
https://inmon.com/technology/sflowlicense.txt

File: doc/sflow/sflow_hld.md (hunk 5 above)
Comment (padmanarayana) [reply doc/sflow/sflow_hld.md]: Collector and agent specific configurations are available in the configDB and host-sflow will have a mod_sonic that is customised to interact with SONiC (essentially query the DB). If there are new customizations, we publish the changes in SONiC schema and mod_sonic will have to be updated (in host-sflow repo) to provide the necessary functionality.

**Comment by kefins** [comment]
Netlink message can not be received by hsflowd, and the psamle driver returns **-ESRCH** while broadcasting message to userspace by `genlmsg_multicast_netns`, can anybody give some advice?

**Comment by prsunny** [comment]
> Netlink message can not be received by hsflowd, and the psamle driver returns **-ESRCH** while broadcasting message to userspace by `genlmsg_multicast_netns`, can anybody give some advice?

Can you please raise an issue to sonic-builidimage? Also with configs/step to reproduce etc.

**Comment by kefins** [comment]
> > Netlink message can not be received by hsflowd, and the psamle driver returns **-ESRCH** while broadcasting message to userspace by `genlmsg_multicast_netns`, can anybody give some advice?
> 
> Can you please raise an issue to sonic-builidimage? Also with configs/step to reproduce etc.

Thank you for replying, finally I solved the problem, which was caused by the network mode of sflow docker. After restarting sflow docker with host network mode, everything went well, thanks again.

**Pull request merged**